### PR TITLE
Feat/aut 3625/content creation mode switch

### PR DESCRIPTION
**Pull request by Nevermind23**
## Task
https://oat-sa.atlassian.net/browse/AUT-3625

## Changelog
- Add interface mode in forms
- Implement new way of accessing/updating user settings via set/get
- Add interface mode in user settings
- Update tests

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -67,7 +67,7 @@
         "oat-sa/jig": "~0.2",
         "oat-sa/composer-npm-bridge": "~0.4.2||dev-master",
         "oat-sa/oatbox-extension-installer": "~1.1||dev-master",
-        "oat-sa/generis": ">=15.34",
+        "oat-sa/generis": "dev-feat/AUT-3625/content-creation-mode-switch",
```
Comment (Nevermind23) [inline composer.json:70]: - [ ] Update generis version

File: actions/class.UserSettings.php [hunk 2]
```diff
@@ -164,6 +175,13 @@ private function getLanguageService(): tao_models_classes_LanguageService
         return tao_models_classes_LanguageService::singleton();
     }
 
+    private function getFeatureFlagChecker(): FeatureFlagChecker
```
Comment (gabrielfs7) [inline actions/class.UserSettings.php:178]: For better open/close principle, better maintainability

```suggestion
    private function getFeatureFlagChecker(): FeatureFlagCheckerInterface
```

File: actions/form/class.UserSettings.php [hunk 3]
```diff
@@ -142,6 +145,22 @@ protected function initElements()
...
+        if (
+            $this->getFeatureFlagChecker()->isEnabled(
+                FeatureFlagCheckerInterface::FEATURE_FLAG_SOLAR_DESIGN_ENABLED
+            )
+        ) {
+            $this->addInterfaceModeEl($this->form);
+        }
+    }
+
+    private function addInterfaceModeEl($form): void
```
Comment (gabrielfs7) [inline actions/form/class.UserSettings.php:158]: Some points, 

- Lets use typehint in the method, please
- I know you are following the naming in place, but we must avoid abbreviation when possible for better reading `addInterfaceModeElement`

File: actions/form/class.UserSettings.php [hunk 4]
```diff
@@ -142,6 +145,22 @@ protected function initElements()
...
+    {
+        $interfaceModeElement = tao_helpers_form_FormFactory::getElement('interface_mode', 'Radiobox');
```
Comment (gabrielfs7) [inline actions/form/class.UserSettings.php:160]: suggestion: Lets use constants for better tracking

```suggestion
        $interfaceModeElement = tao_helpers_form_FormFactory::getElement(UserSettingsInterface::INTERFACE_MODE, 'Radiobox');
```

File: models/classes/menu/SectionVisibilityFilter.php [hunk 5]
```diff
@@ -23,16 +23,21 @@
...
+use oat\tao\model\user\implementation\UserSettingsService;
+use oat\tao\model\user\UserSettingsInterface;
 
 class SectionVisibilityFilter extends ConfigurableService implements SectionVisibilityFilterInterface
 {
     public const SERVICE_ID = 'tao/SectionVisibilityFilter';
     public const OPTION_FEATURE_FLAG_SECTIONS = 'featureFlagSections';
     public const OPTION_FEATURE_FLAG_SECTIONS_TO_HIDE = 'featureFlagSectionsToHide';
 
+    //TODO:: Add sections hidden for simple mode
```
Comment (gabrielfs7) [inline models/classes/menu/SectionVisibilityFilter.php:39]: Is this TODO still actionable?

File: models/classes/menu/SectionVisibilityFilter.php [hunk 6]
```diff
@@ -23,16 +23,21 @@
...
+    public const SIMPLE_MODE_HIDDEN_SECTIONS = [];
```
Comment (gabrielfs7) [inline models/classes/menu/SectionVisibilityFilter.php:40]: We must separate constants from other sections by black line.
```suggestion
    public const SIMPLE_MODE_HIDDEN_SECTIONS = [];
    
```

File: models/classes/menu/SectionVisibilityFilter.php [hunk 7]
```diff
@@ -53,10 +58,29 @@ public function isVisible(string $section): bool
             }
         }
 
+        $userSettings = $this->getUserSettingsService()->getCurrentUserSettings();
+        if (
```
Comment (gabrielfs7) [inline models/classes/menu/SectionVisibilityFilter.php:62]: Blocks are separate by blank lines in OAT php standards
```suggestion
        $userSettings = $this->getUserSettingsService()->getCurrentUserSettings();
        
        if (
```

Comment (gabrielfs7) [reply models/classes/menu/SectionVisibilityFilter.php:62]: Another point, lets avoid getting $userSettings, unless the feature flag is enabled. Better to move this condition to inside the IF

File: models/classes/user/UserSettingsInterface.php [hunk 8]
```diff
@@ -24,9 +24,15 @@
 
 interface UserSettingsInterface
 {
+    public const INTERFACE_MODE = 'interface_mode';
```
Comment (gabrielfs7) [inline models/classes/user/UserSettingsInterface.php:27]: Lets please add all the settings as constant, so we can start using only the `getSetting` method to get them and remove any temptation of developers to extend UserSettingsInterface with more custom methods per parameters, which is not sustainable for this entity and causes break changes.

Lets please `@deprecate` all other methods except for `get/setSetting`

```suggestion
    public const INTERFACE_MODE = 'interface_mode';
    public const UI_LANGUAGE_CODE = 'ui_language_code';
    public const DATA_LANGUAGE_CODE = 'data_language_code';
    public const TIMEZONE = 'timezone';
```

File: models/classes/user/implementation/UserSettingsService.php [hunk 9]
```diff
@@ -65,10 +75,29 @@ public function get(core_kernel_classes_Resource $user): UserSettingsInterface
...
         }
 
-        return new UserSettings(
+        $userSettings = new UserSettings(
             $timezone ?? $this->defaultTimeZone,
             $uiLanguageCode ?? null,
             $dataLanguageCode ?? null
         );
+
+        if (!empty($props[GenerisRdf::PROPERTY_USER_INTERFACE_MODE])) {
```
Comment (gabrielfs7) [inline models/classes/user/implementation/UserSettingsService.php:84]: We also need to check the feature flag here, cause we may have disabled the feature flag after have saved this config

File: models/classes/user/implementation/UserSettingsService.php [hunk 10]
```diff
@@ -65,10 +75,29 @@ public function get(core_kernel_classes_Resource $user): UserSettingsInterface
...
+    }
+
+    public function getCurrentUserSettings(): UserSettingsInterface
+    {
+        $currentUser = $this->userService->getCurrentUser();
+        if ($currentUser) {
+            return $this->get($currentUser);
+        }
+
+        return new UserSettings($this->defaultTimeZone);
```
Comment (gabrielfs7) [inline models/classes/user/implementation/UserSettingsService.php:101]: Please, never glue `if` with other `ifs` or other instructions.

```suggestion
        $currentUser = $this->userService->getCurrentUser();
        
        if ($currentUser) {
            return $this->get($currentUser);
        }

        return new UserSettings($this->defaultTimeZone);
```

File: test/unit/models/classes/menu/SectionVisibilityFilterTest.php [hunk 11]
```diff
@@ -37,9 +38,13 @@ class SectionVisibilityFilterTest extends TestCase
     /** @var FeatureFlagChecker|MockObject  */
     private $featureFlagChecker;
 
+    /** @var FeatureFlagChecker|MockObject  */
```
Comment (gabrielfs7) [inline test/unit/models/classes/menu/SectionVisibilityFilterTest.php:41]:
```suggestion
    /** @var UserSettingsServiceInterface|MockObject  */
```

File: test/unit/user/implementation/UserSettingsServiceTest.php [hunk 12]
```diff
@@ -134,7 +157,8 @@ private function getUserMock(
             ->with([
                 $this->getOntologyMock()->getProperty(GenerisRdf::PROPERTY_USER_UILG),
                 $this->getOntologyMock()->getProperty(GenerisRdf::PROPERTY_USER_DEFLG),
-                $this->getOntologyMock()->getProperty(GenerisRdf::PROPERTY_USER_TIMEZONE)
+                $this->getOntologyMock()->getProperty(GenerisRdf::PROPERTY_USER_TIMEZONE),
+                $this->getOntologyMock()->getProperty(GenerisRdf::PROPERTY_USER_INTERFACE_MODE)
```
Comment (gabrielfs7) [inline test/unit/user/implementation/UserSettingsServiceTest.php:161]: Can we also test this? We added it here, but we do not test getting this property

File: actions/class.UserSettings.php [hunk 13]
```diff
@@ -110,6 +112,15 @@ public function properties()
                 $userSettingsData[GenerisRdf::PROPERTY_USER_DEFLG] = $dataLang->getUri();
             }
 
+            if (
+                $this->getFeatureFlagChecker()->isEnabled(
+                    FeatureFlagCheckerInterface::FEATURE_FLAG_SOLAR_DESIGN_ENABLED
+                )
+            ) {
+                $interfaceMode = $this->getResource($settingsForm->getValue('interface_mode'));
```
[the remark this answers is not in the document]
Comment (Nevermind23) [reply actions/class.UserSettings.php:120]: The one from interface is for the DB, the other one is for FE form name, is this fine if we use same constant?

File: models/classes/menu/SectionVisibilityFilter.php (hunk 5 above)
Comment (Nevermind23) [reply models/classes/menu/SectionVisibilityFilter.php:39]: I am not sure if we have a list of sections that should be hidden in simple interface mode.

File: models/classes/user/implementation/UserSettingsService.php (hunk 9 above)
Comment (Nevermind23) [reply models/classes/user/implementation/UserSettingsService.php:84]: This is for getting the value of the interface mode, no matter if feature flag is disabled or enabled, I was always getting the value of the interface mode.
But I was always checking for feature flag, whenever this value is used.

Do you think we should return empty string if the feature flag is not set?

File: actions/class.UserSettings.php (hunk 13 above)
[the remark this answers is not in the document]
Comment (gabrielfs7) [reply actions/class.UserSettings.php:120]: I would say, it is better, cause we keep consistency.

File: models/classes/menu/SectionVisibilityFilter.php (hunk 5 above)
Comment (gabrielfs7) [reply models/classes/menu/SectionVisibilityFilter.php:39]: Ok, lets then confirm with @shaveko and close this TODO after we confirm all

File: models/classes/user/implementation/UserSettingsService.php (hunk 9 above)
Comment (gabrielfs7) [reply models/classes/user/implementation/UserSettingsService.php:84]: As commented in other PRs, lets populate this value (or not, make it null) also based on the feature flag, so you do not need to keep asking this feature flag on all the places you read this config? So we centralize the responsibility here, which is the right place. The persistence layer is just an implementation detail, but the client of this class does not need to care about it

File: models/classes/user/UserSettingsInterface.php (hunk 8 above)
Comment (gabrielfs7) [reply models/classes/user/UserSettingsInterface.php:27]: Thanks @Nevermind23 for the fix!

File: models/classes/user/implementation/UserSettingsService.php (hunk 9 above)
Comment (gabrielfs7) [reply models/classes/user/implementation/UserSettingsService.php]: Thanks @Nevermind23 for the fix!

**Review by gabrielfs7 — APPROVED** [review]
Awesome work, Thanks @Nevermind23 for the fix!

- [x] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful

File: models/classes/action/ActionBlackList.php [hunk 14]
```diff
@@ -64,4 +72,28 @@ private function getFeatureFlagChecker(): FeatureFlagChecker
...
+
+        if (
+            $userSettings->getSetting(
+                UserSettingsInterface::INTERFACE_MODE
+            ) === GenerisRdf::PROPERTY_USER_INTERFACE_MODE_SIMPLE
+        ) {
+            return true;
+        }
+
+        return false;
```
Comment (gabrielfs7) [inline models/classes/action/ActionBlackList.php:97]: chore: can be simplified:

```suggestion
        return $this->getUserSettingsService()
            ->getCurrentUserSettings()
            ->getSetting(UserSettingsInterface::INTERFACE_MODE) === GenerisRdf::PROPERTY_USER_INTERFACE_MODE_SIMPLE
        );
```

**Review by gabrielfs7 — APPROVED** [review]

File: test/unit/user/implementation/UserSettingsServiceTest.php [hunk 15]
```diff
@@ -74,36 +87,126 @@ public function testGet(
...
+        string $userInterfaceMode = null
+    ): UserSettingsInterface {
+        $userSettings = new UserSettings($userTimezone);
+
+        $userSettings->setSetting(UserSettingsInterface::INTERFACE_MODE, $userInterfaceMode);
+        $userSettings->setSetting(UserSettingsInterface::UI_LANGUAGE_CODE, $uiLanguageUri);
+        $userSettings->setSetting(UserSettingsInterface::DATA_LANGUAGE_CODE, $defLangUri);
+        $userSettings->setSetting(UserSettingsInterface::TIMEZONE, $userTimezone);
+
+        return $userSettings;
```
Comment (gabrielfs7) [inline test/unit/user/implementation/UserSettingsServiceTest.php:202]: chore: Can be simplified with fluent interface

```suggestion
        return $userSettings->setSetting(UserSettingsInterface::INTERFACE_MODE, $userInterfaceMode)
            ->setSetting(UserSettingsInterface::UI_LANGUAGE_CODE, $uiLanguageUri)
            ->setSetting(UserSettingsInterface::DATA_LANGUAGE_CODE, $defLangUri)
            ->setSetting(UserSettingsInterface::TIMEZONE, $userTimezone);
```

**Pull request merged**